### ISSUE TITLE
Waiting for Events in the Command Buffers Extension

### ISSUE THREAD

**Issue by FreddieWitherden** [KhronosGroup/OpenCL-Docs#793]
I've been going through the proposed command buffer extension, to see if our application, PyFR, might benefit from it.

However, while attempting to port the OpenCL backend of our application over, I encountered an issue around events.  Specifically, that it does not appear to be possible to wait on any of the kernels inside of a command buffer.  This functionality is extremely useful in the context of HPC where it is necessary to interact with external libraries such as
MPI.  The idea is to have kernels which pack buffers and copy them to the host and then, as they complete, have the host-side kick off the relevant MPI_Isend calls.  This is easy with regular OpenCL using events, and with CUDA/HIP's graph APIs using their event record nodes.

One solution here might be to add a

```c
cl_int clCreateCommandSyncEventKHR(
   cl_command_buffer_khr command_buffer,
   cl_sync_point_khr* sync_point,
   cl_uint *ix
);
```
method to convert a sync point to an index and then a:
```c
cl_int clEnqueueCommandBufferWithSyncEventsKHR(
     cl_uint num_queues,
     cl_command_queue* queues,
     cl_command_buffer_khr command_buffer,
     cl_uint num_sync_events,
     cl_event* sync_event_list,
     cl_uint num_events_in_wait_list,
     const cl_event* event_wait_list,
     cl_event* event);
```
which will mint a set of cl_events with the number of events being equal to the number of times clCreateCommandSyncEventKHR was called during recording.  On platforms without any kind of native or preferred support for host side waiting, the functionality can be emulated internally without too much hassle (by cracking apart the command buffer into multiple smaller command buffers in an implementation defined manner).

**Comment by FreddieWitherden** [comment]
Yes, my particular need is only for device -> host synchronization.  However, the other direction can also be supported easily enough via:
```c
cl_sync_point *clCreateCommandUserSyncPointKHR(
   cl_command_buffer_khr command_buffer,
   cl_uint *ix, c_int *err
);
```
or something to that effect which creates a new sync point which can be used within the context of the command buffer.  When `clEnqueueCommandBufferWithSyncEventsKHR` is called some of the returned events will be user events.

**Comment by bashbaug** [comment]
This is a really great question!

One of the nice properties (at least, from an implementor's perspective) of the current command buffer extension is that a command buffer is entirely "self-contained", meaning the commands within a command buffer can wait on other commands in a command buffer, but they cannot wait on events from outside of the command buffer, and commands outside of a command buffer cannot wait on a command buffer sync point.  This gives an implementation a great deal of optimization freedom and avoids tricky issues like event lifetimes, though it is admittedly a little restrictive.

A few other solutions we could consider, in addition to the ones already described in this issue:

* Could we allow signaling an [OpenCL semaphore](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#cl_khr_semaphore) from a command buffer?  This would most likely be useful for device-to-device synchronization while a command buffer is executing, but it could also be coupled with a wait on semaphore and an event callback (in a different queue?) to get the device-to-host behavior.
    * Note, this would probably only work for a binary semaphore, or we would need a way to update the payload when signaling a timeline semaphore.
    * Additonal note, we could allow waiting on a semaphore from a command buffer also, though this is a little trickier and could lead to deadlock, so we should treat signaling and waiting as two separate features.

* Could implementations that support [SVM](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#shared-virtual-memory) or [USM](https://www.khronos.org/registry/OpenCL/extensions/intel/cl_intel_unified_shared_memory.html) write to a host-accessible memory address to trivially provide device-to-host synchronization?  Would doing so require any memory model considerations to guarantee that if the device-to-host signal is observed then any previous buffer writes are observable also?

**Comment by EwanC** [comment]
Noting on this ticket that we have been looking into supporting something similar with SYCL-Graph - https://github.com/intel/llvm/pull/15056 We haven't implemented that yet, but have prototyped support at the [unified runtime level](https://oneapi-src.github.io/unified-runtime/core/EXP-COMMAND-BUFFER.html#command-synchronization-with-events). Where extra parameters have been added to command append entry-points for the ur events to wait on and signal. This API is [only implemented for CUDA](https://github.com/oneapi-src/unified-runtime/pull/1823) where we add `cuGraphEventRecordNodeGetEvent` nodes to the graph for the output ur event, and `cuGraphAddEventWaitNode` nodes for the input `ur` event.
 
I'm not wedded to this UR API, and if OpenCL decided to go for a different approach from signalling to/from inside a command-buffer then it could be updated to be compatible. But more mentioning it as a point of reference for how this functionality could be taken advantage of up the stack. Being able to update the cl events signalled and waited should also be a consideration.
